dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1341,85 +1577,233 @@ RtpsUdpDataLink::send_ack_nacks(RtpsReaderMap::value_type& rr, bool finalFlag)
+        accumulate_addresses(it->from_guid_, *it2, addrs);
+      }
+      if (addrs.empty()) {
+        continue;
+      }
+      if (std::memcmp(&(it->dst_guid_.guidPrefix), &GUIDPREFIX_UNKNOWN, sizeof(GuidPrefix_t)) != 0) {
+        RepoId dst;
+        memcpy(dst.guidPrefix, it->dst_guid_.guidPrefix, sizeof(dst.guidPrefix));
+        dst.entityId = ENTITYID_UNKNOWN;
+        adr_map[addrs][dst].push_back(it);
+      } else {
+        adr_map[addrs][GUID_UNKNOWN].push_back(it);
+      }
+      addrs.clear();
+    }
+  }
+
+  // Reusable INFO_DST
+  InfoDestinationSubmessage idst = {
+    {INFO_DST, FLAG_E, INFO_DST_SZ},
+    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+  };
+
+  // Build bundles
+  typedef OPENDDS_VECTOR(ResponseIterVec) ResponseIterVecVec;
+  ResponseIterVecVec response_bundles; // a vector of vectors of iterators pointing to responses
+  OPENDDS_VECTOR(AddrSet) response_bundle_addrs; // for a bundle's address set
+  OPENDDS_VECTOR(size_t) response_bundles_sizes; // for allocating the bundle's buffer
+  size_t size = 0, padding = 0, prev_size = 0, prev_padding = 0; // keep track of previous values in case we overshoot
+  RepoId prev_dst; // used to determine when we need to write a new info_dst
+  for (AddrDestResponseMap::iterator addr_it = adr_map.begin(); addr_it != adr_map.end(); ++addr_it) {
+
+    // A new address set always starts a new bundle
+    response_bundles.push_back(ResponseIterVec());
+    response_bundle_addrs.push_back(addr_it->first);
+    if (addr_it != adr_map.begin()) {
+      response_bundles_sizes.push_back(prev_size + prev_padding);
+      size -= prev_size; padding -= prev_padding; prev_size = 0; prev_padding = 0;
+    }
+    prev_dst = GUID_UNKNOWN;
+
+    for (DestResponseMap::iterator dest_it = addr_it->second.begin(); dest_it != addr_it->second.end(); ++dest_it) {
+      size_t offset = size;
+      for (ResponseIterVec::iterator resp_it = dest_it->second.begin(); resp_it != dest_it->second.end(); ++resp_it) {
+        // Check before every response to see if we need to prefix a INFO_DST
+        if (dest_it->first != GUID_UNKNOWN && dest_it->first != prev_dst) {
+          gen_find_size(idst, size, padding);
+          prev_dst = dest_it->first;
+        }
+        // If adding an INFO_DST prefix bumped us over the limit, push the size difference into the next bundle, reset prev_dst, and keep going
+        if ((size + padding) > max_bundle_size_) {
+          response_bundles.push_back(ResponseIterVec());
+          response_bundle_addrs.push_back(addr_it->first);
+          response_bundles_sizes.push_back(prev_size + prev_padding);
+          size -= prev_size; padding -= prev_padding; prev_size = 0; prev_padding = 0;
+          prev_dst = GUID_UNKNOWN;
+        } else {
+          prev_size = size; prev_padding = padding;
+        }
+        // Attempt to add the submessage response to the bundle
+        offset = size;
+        Response& res = **resp_it;
+        switch (res.sm_._d()) {
```

Comment:
Refactor this to its own function?  (This, with the other changes, may get the resulting `send_bundled_responses` within range of the arbitrary "fits on one page" metric.)